src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -906,6 +907,81 @@ static bool usesReceiver(BoundExpression argument)
             }
         }
 
+        [return: NotNullIfNotNull(nameof(receiver))]
+        BoundExpression? AdjustReceiverForExtensionsIfNeeded(BoundExpression? receiver, Symbol member)
+        {
+            if (receiver is null)
+            {
+                return receiver;
+            }
+
+            Debug.Assert(member is not null);
+            if (receiver is BoundTypeExpression)
+            {
+                return receiver;
+            }
+
+            if (receiver is BoundThisReference)
+            {
+                // PROTOTYPE this was only minimally tested to observe side-effects but needs more work (in CheckValue for instance)
+                // In an extension type, `this.UnderlyingTypeMember` is replaced with `this.<UnderlyingInstanceField>$.UnderlyingTypeMember`
+                Debug.Assert(receiver.Type is not null);
+                if (receiver.Type.IsExtension && !member.ContainingType.IsExtension)
+                {
+                    var thisExtensionType = (SourceExtensionTypeSymbol)receiver.Type;
+                    Debug.Assert(thisExtensionType.UnderlyingInstanceField is not null);
+                    Debug.Assert(thisExtensionType.UnderlyingInstanceField.Type.Equals(member.ContainingType, TypeCompareKind.AllIgnoreOptions));
+
+                    return _factory.Field(receiver, thisExtensionType.UnderlyingInstanceField);
+                }
+            }
+
+            if (member.ContainingType is not { IsExtension: true } memberExtensionType)
+            {
+                return receiver;
+            }
+
+            Debug.Assert(receiver.Type is not null);
+            if (receiver.Type.IsExtension)
+            {
+                return receiver;
+            }
+
+            ArrayBuilder<LocalSymbol>? temps = ArrayBuilder<LocalSymbol>.GetInstance();
+            ArrayBuilder<BoundExpression>? effects = ArrayBuilder<BoundExpression>.GetInstance();
+            if (receiver.Type.IsReferenceType)
+            {
+                receiver = MakeTemp(receiver, temps, effects);
+            }
+            else if (receiver.Type.IsValueType)
+            {
+                receiver = MakeTemp(receiver, temps, effects, RefKind.Ref);
+            }
+            else
+            {
+                throw ExceptionUtilities.UnexpectedValue(receiver.Type);
+            }
+
+            Debug.Assert(receiver.Type is not null);
+            var oldSyntax = _factory.Syntax;
+            _factory.Syntax = receiver.Syntax;
+
+            var call = _factory.Call(null,
+                            _factory.WellKnownMethod(WellKnownMember.System_Runtime_CompilerServices_Unsafe__As_T)
```

Comment:
>System_Runtime_CompilerServices_Unsafe__As_T [](http://example.com/codeflow?start=69&length=44)

We need to verify that we can rewrite this call across `await`. I.e. it should be possible to `await` in the argument list when we do this transform. Please make sure to add relevant tests.

Reply:
Added `AdjustReceiver_AwaitArgument`. It errors. I left a PROTOTYPE comment as I didn't investigate closely whether we could make it work